### PR TITLE
rmfakecloud: init package and module

### PR DESCRIPTION
**Pull request by pacien**
###### Motivation for this change

This adds a package and a module for [rmfakecloud], an open-source
implementation of the cloud synchronisation server for the Remarkable e-ink
tablets.

[rmfakecloud]: https://ddvk.github.io/rmfakecloud/

The package only includes the server itself, without the JS web UI, because the
latter is hard to build on NixOS. The server remains nevertheless perfectly
usable without it thanks to its command-line interface.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (roosemberth) [inline nixos/modules/services/misc/rmfakecloud.nix:110]: The following is solely opinion-based of what I'd like to see here. AFAIK, nixpkgs has no defined threat model modules should align themselves to nor a policy on recommended security features modules should use.

```suggestion
        ProtectControlGroups = true;
        CapabilityBoundingSet = [ "" ];
        DevicePolicy = "closed";
        LockPersonality = true;
        MemoryDenyWriteExecute = true;
        ProtectClock = true;
        ProtectHostname = true;
        ProtectKernelLogs = true;
        ProtectProc = "invisible";
        ProcSubset = "pid";
        RemoveIPC = true;
        RestrictAddressFamilies = [
          "AF_INET"
          "AF_INET6"
        ];
        RestrictNamespaces = true;
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
        SystemCallArchitectures = "native";
```

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (roosemberth) [inline nixos/modules/services/misc/rmfakecloud.nix:88]: I think the package used by the module should be an overridable option: The package proposed here does not include the user SPA (which is fine), but shall users of this module want to build it with the app, they should be able to write such derivation and have this module use it.

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (roosemberth) [inline nixos/modules/services/misc/rmfakecloud.nix:79]: I'm a bit troubled by this since `JWT_SECRET_KEY` may have been defined in the environment file, thus making this code not only rather unuseful but also confusing by creating an unused secret.

Moreover, the extra dependency on openssl pushes an additional 90MB to the closure (it could be argued that openssl may already be in the system, but I'm still reticent to add it here).

I see two acceptable solutions:

- Add an option `autoCreateJwtSecret`, defaults to true and informs the user that if such secret does not exist and is not defined in the envvars, it shall be created. If set to false, don't bring openssl into the closure but merely assert that either the file exists or the envvar is defined.
- Check whether the envvar is defined before creating the secret. I'm not sure whether preStart has access to the environment (my guess it that it does).

**Review by roosemberth — CHANGES_REQUESTED** [review]
Thanks for your contribution. I've left some remarks, but I'm super excited about this!

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix]: Those description were direct copy-pastes from the original documentation of the software, but I do prefer your more succinct proposals.

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix:55]: Those `environmentFile" options seem to be fairly common in nixpkgs. Do you know whether there's some generic documentation template that should be used here instead? Or even better, some section in a manual that could be referenced here?

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix:86]: The `rmfakecloud` package should eventually be completed to include the web UI as well. I do not think that there is any reason to propose different variants of the package, so I do not think that a `package` option in the module would be useful.

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (roosemberth) [reply nixos/modules/services/misc/rmfakecloud.nix:55]: Not that I know of, unfortunately :(

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (roosemberth) [reply nixos/modules/services/misc/rmfakecloud.nix:86]: I think some users may still prefer to have the package without the SPA (myself included), so imho, the package should eventually include an argument to choose whether the SPA should be compiled. It follows that an option package would probably be useful.

Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix]: Okay, I added a `package` option, since this does not cost anything.

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix]: Agreed. I implemented your second suggestion. I also removed the dependency over `openssl` for generating the random secret.

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (pacien) [reply nixos/modules/services/misc/rmfakecloud.nix:7]: Other modules seem to be dropping options to set custom data directories because they are hard to manage when `DynamicUser` is used for the systemd service.

The effective value can still be overridden by mounting another volume or directory to `/var/lib/private/rmfakecloud`, by altering the systemd service's option, and/or by setting `extraSettings.DATADIR` to some other path. The user is then responsible to deal with potential permission issues themselves.

Comment (roosemberth) [reply nixos/modules/services/misc/rmfakecloud.nix:7]: > Other modules seem to be dropping options to set custom data directories because they are hard to manage when DynamicUser is used for the systemd service.

Yes, you are right, DynamicUser is problematic with existing directories...

I still find this annoying from the point that if `DATADIR` were to be set, the service would still create a useless `/var/lib/rmfakecloud` that will only be confusing. It doesn't bother me enough to block the MR because of that...

**Review by roosemberth — APPROVED** [review]
Great work! Thank you!

File: pkgs/servers/rmfakecloud/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/rmfakecloud/default.nix:22]:
```suggestion
    "-s" "-w" "-X main.version=v${version}"
```

Comment (pacien) [reply pkgs/servers/rmfakecloud/default.nix]: Suggestion amended, thanks!

File: pkgs/servers/rmfakecloud/default.nix
Comment (martinetd) [inline pkgs/servers/rmfakecloud/default.nix:17]: For what it's worth I was able to build it with the JS thingie:
http://cgit.notk.org/asmadeus/nixos-config.git/tree/pkgs/rmfakecloud/default.nix

I don't use it though so don't particularly care and will switch to this package regardless on this gets in, but it shouldn't be much different.

File: pkgs/servers/rmfakecloud/default.nix
Comment (martinetd) [inline pkgs/servers/rmfakecloud/default.nix:5]: (hindsight 20/20: there's been a v0.0.6 release a few days ago, probably better to use than some random git rev; happy to upgrade after this gets in though)

File: nixos/modules/services/misc/rmfakecloud.nix
Comment (martinetd) [reply nixos/modules/services/misc/rmfakecloud.nix:123]: For a network-facing service I'd go as far as suggest using `systemd.services.rmfakecloud.confinement` which works well for this service: http://cgit.notk.org/asmadeus/nixos-config.git/tree/profiles/rmfakecloud.nix
(But happy to keep it locally though, it doesn't like DynamicUser...)

File: pkgs/servers/rmfakecloud/default.nix
Comment (martinetd) [inline pkgs/servers/rmfakecloud/default.nix:3]: two notes:
 - there's been a arms-long list of discussions about buildGoModules vs. buildGoPackage because buildGoModules duplicates all go dependencies/fixed output derivations are frowned upon, see https://github.com/NixOS/nixpkgs/issues/84826 
That being said, I don't think we should block on this as the new shiny solution (gomod2nix) isn't merged in nixpkgs yet
 - if you update to v0.0.6 (see next comment) you might need to switch to buildGo117Modules, I wasn't able to build  the latest release with go 1.16, but that might be related to the UI that I include in my local package

**Review by martinetd — APPROVED** [review]
damn, I wish I had seen this sooner (or this had been merged last week!), I've just finally taken the time to deal with this mess of javascript UI in my repo and was considering pushing the package...! Thanks for the work :)

Anyway, I've:
 - checked the package builds, the workaround is similar to what I used to do so I expect it to work fine
 - reviewed the module & package, looks good to me.

I've added a bunch of comments but that can be done later imo. Also happy if you want to add me (martinetd) to maintainers of the package, but if you're fine alone I don't care either way.

File: pkgs/servers/rmfakecloud/default.nix
Comment (pacien) [reply pkgs/servers/rmfakecloud/default.nix]: There's even a 0.0.7 now! I amended that to this PR.

File: pkgs/servers/rmfakecloud/default.nix
Comment (pacien) [reply pkgs/servers/rmfakecloud/default.nix:3]: I'm not sure of how to migrate to `buildGoPackage`. If it's still acceptable until all the tools are merged, let's keep it that way for now.

The package seems to build just fine with the default version of the channel (1.16) for me (without the web UI for now).

File: pkgs/servers/rmfakecloud/default.nix
Comment (pacien) [reply pkgs/servers/rmfakecloud/default.nix:17]: Thanks! I think this should be made in a separate subsequent PR to this one.

Comment (martinetd) [reply pkgs/servers/rmfakecloud/default.nix:17]: Definitely agree there, this is already great -- no reason to hold this PR back :)
I'll be happy to submit the UI if it turns out it's useful, I actually haven't even looked at it yet...

File: pkgs/servers/rmfakecloud/default.nix
Comment (martinetd) [reply pkgs/servers/rmfakecloud/default.nix]: Cheers, can confirm new version also builds.

**Review by martinetd — APPROVED** [review]
(re-approving for ofborg, I assume this is because I was added to maintainers)

**Pull request merged**
